server/src/main/java/keywhiz/service/daos/SecretController.java

```diff
@@ -348,5 +378,21 @@ public Secret createOrUpdate() {
           generationOptions);
       return transformer.transform(secretDAO.getSecretByName(name).get());
     }
+
+    public Secret createOrUpdate(Object principal) {
+      secretDAO.createOrUpdateSecret(
+          name,
+          ownerName,
+          encryptedSecret,
+          hmac,
+          creator,
+          metadata,
+          expiry,
+          description,
+          type,
+          generationOptions,
+          principal);
+      return transformer.transform(secretDAO.getSecretByName(name, principal).get());
+    }
```

Comment:
Having the business logic in controller does make sense, although there are some methods in SecretResource like partialUpdateSecret which call the DAO directly instead of going through the controller. Maybe in the future we could change all the methods to call the controller instead.

Reply:
Yeah, and controller is a very MVC-ish term (although I think it actually works here), but typically I like to put business logic for Things into ThingManager. I know a lot of people don't like the "Manager" name, but it's obvious and it saves you the hassle of trying to come up with something more clever.